enhancements/console/generated-resource-types.md

```diff
@@ -0,0 +1,217 @@
+
+* Periodically publish a package on npm which contains TypeScript types for k8s
+  resources used by the web console, based on their OpenAPI specifications.
+* Consume this package in the web console and in associated console dynamic plugins.
+
+### Non-Goals
+
+This enhancement will not seek to implement:
+
+* Validation of objects against the generated types during runtime.
+* Generation of types for all CustomResourceDefinitions used by the web console.
+  The initial focus will be on core k8s resources and the CRDs defined by the
+  `openshift/api` repository.
+
+## Proposal
+
+### Workflow Description
+
+**library authors** are responsible for maintaining the type generation library.
+1. They will run the generation scripts at some point during the OpenShift release
+2. The generated types will be published to a package on npm (e.g. `@openshift/k8s-types`)
+
+**web console developers** will consume the generated types in the web console.
+1. They will add or update the dependency to the package in the web console's `package.json`.
+2. They will import the types from the package in the web console's codebase.
+3. They will fix resulting type errors in the web console's codebase due to
+   changes in the generated types.
+
+
+### API Extensions
+NA
+
+### Risks and Mitigations
+
+The proposal aims to mitigate risks which already exist.
+
+### Drawbacks
+
+* Web console developers do not have unlimited control over types which are used
+  in the codebase.
+
+## Test Plan
+
+As this enhancement proposal focuses on the generation of TypeScript types,
+existing unit and integration tests that utilize these types will serve as the
+primary means of validation. To ensure the correctness of the generated types,
+we will implement the following test strategies:
+
+* **Unit tests in the generation library**: We will add unit tests to the type
+  generation library to verify that it correctly transforms OpenAPI specifications
+  into TypeScript types.
+* **Integration tests in the web console**: We will leverage existing unit and
+  integration tests in the web console that utilize the generated types. These
+  tests will help ensure that the generated types are accurate and function as
+  expected within the context of the web console.
+* **Manual verification**: We will perform manual verification of the generated
+  types by inspecting them and comparing them to the original OpenAPI definitions.
+
+## Graduation Criteria
+N/A
+
+### Dev Preview -> Tech Preview
+N/A
+
+### Tech Preview -> GA
+N/A
+
+### Removing a deprecated feature
+N/A
+
+## Upgrade / Downgrade Strategy
+N/A
+
+## Version Skew Strategy
+
+Version skew is mitigated by the fact that resources are not frequently updated
+in a way that would break existing types. An older version of the types package
+should still be compatible with newer versions of the resources, as long as the
+resources have not undergone breaking changes which would impact the web console's
+usage of them.
```

Comment:
It's also mitigated by OpenShift policy of not breaking APIs. Presumably we will publish different types for different API versions?